lib/Basics/debugging.h

```diff
@@ -199,31 +200,54 @@ enable_if_t<is_container<T>::value, std::ostream&> operator<<(std::ostream& o, T
   return o;
 }
 
+namespace debug {
+struct NoOpStream {
+  template <typename T>
+  auto operator<<(T const&) -> NoOpStream& {
+    return *this;
+  }
+};
+
+struct AssertionLogger {
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+  void operator&(std::ostringstream const& stream) const {
+    std::string message = stream.str();
+    arangodb::CrashHandler::assertionFailure(file, line, function, expr,
+                                             message.empty() ? nullptr : message.c_str());
+  }
+  // can be removed in C++20 because of LWG 1203
+  void operator&(std::ostream const& stream) const {
+    operator&(static_cast<std::ostringstream const&>(stream));
+  }
+#endif
+  void operator&(NoOpStream const&) {}
+  const char* file;
+  int line;
+  const char* function;
+  const char* expr;
+};
+
+}  // namespace debug
+
 }  // namespace arangodb
 
 /// @brief assert
 #ifndef TRI_ASSERT
 
 #ifdef ARANGODB_ENABLE_MAINTAINER_MODE
 
-#define TRI_ASSERT(expr) /*GCOVR_EXCL_LINE*/                                             \
-  do {                                                                                   \
-    if (!(ADB_LIKELY(expr))) {                                                           \
-      arangodb::CrashHandler::assertionFailure(__FILE__, __LINE__, __FUNCTION__, #expr); \
-    } else {                                                                             \
-    }                                                                                    \
-  } while (false)
+#define TRI_ASSERT(expr) /*GCOVR_EXCL_LINE*/                                          \
+  (ADB_LIKELY(expr))                                                                  \
+      ? (void)nullptr                                                                 \
+      : ::arangodb::debug::AssertionLogger{__FILE__, __LINE__, __FUNCTION__, #expr} & \
+            std::ostringstream {}
 
 #else
 
-#define TRI_ASSERT(expr) /*GCOVR_EXCL_LINE*/ \
-  do {                                       \
-    if (false) {                             \
-      (void)(expr);                          \
-    }                                        \
-  } while (false)
+#define TRI_ASSERT(expr) /*GCOVR_EXCL_LINE*/        \
+  (true) ? ((false) ? (void)(expr) : (void)nullptr) \
+         : ::arangodb::debug::AssertionLogger{} & ::arangodb::debug::NoOpStream {}
```

Comment:
Just wondering why we need to create a default-constructed AssertionLogger here? Can't we get away without it in non-maintainer mode?

Reply:
To be clear, we are talking about dead code here. It is there, because we need to cast to void, but a void cast `(void)` has a higher precedence than `<<`. Thus this is done using `operator&`. Exactly as we do in the logger macros.

Reply:
Do we need `epxr` here? I know that the old code also used it, but is it necessary? The problem is that even though the expression is never evaluated, it must at least compile. I only just had a case where I had a member that was only available in maintainer mode which was used in an assertion. So I had to wrap the assertion in yet another ifdef because the compiler would complain that it does not know the member used in the expression.

Reply:
Yes, we need `expr` here!
The reason is that some variables may only be used in assertions, and the compiler will complain about unused variables if the assertion macro just swallows their (sole) usage.

Reply:
Makes sense (unfortunately).